landoapi/secapproval.py

```diff
@@ -0,0 +1,100 @@
+
+```
+{message}
+```
+"""
+
+
+def send_sanitized_commit_message_for_review(revision_phid, message, phabclient):
+    """Send a sanitized commit message for review by the sec-approval team.
+
+    See https://wiki.mozilla.org/Security/Bug_Approval_Process.
+
+    Args:
+        revision_phid: The PHID of the revision to edit.
+        message: The sanitized commit message string we want to be reviewed.
+        phabclient: A PhabClient instance.
+
+    Returns:
+        A list of Phabricator transactions that were generated by the
+        sec-approval request.
+    """
+    comment = SECURE_COMMENT_TEMPLATE.format(message=message)
+    sec_approval_phid = get_sec_approval_project_phid(phabclient)
+    response = phabclient.call_conduit(
+        "differential.revision.edit",
+        objectIdentifier=revision_phid,
+        transactions=[
+            # The caller's alternative commit message is published as a comment.
+            {"type": "comment", "value": comment},
+            # We must get one of the sec-approval project members to approve the
+            # alternate commit message for the review to proceed.
+            #
+            # We need to handle the case where the sec-approval team has approved a
+            # previous alternate commit message and the author has sent a new
+            # message. In this case we want the review to block on the sec-approval
+            # team again and for the sec-approval team to review the new message.
+            # Explicitly re-adding the sec-approval team to the review will clear any
+            # previous reviews the team did and change their state to "blocking,
+            # needs review".  Other reviewers' reviews will be left untouched. The
+            # overall state of the revision will become "Needs Review".
+            #
+            # NOTE: the 'blocking(PHID)' syntax is undocumented at the time of
+            # writing.
+            {"type": "reviewers.add", "value": [f"blocking({sec_approval_phid})"]},
+        ],
+    )
+    return PhabricatorClient.expect(response, "transactions")
+
+
+def save_sec_approval_request_event(revision_phid, diff_phid, transactions):
+    """Save information about a sec-approval review request to the database.
+
+    Args:
+        revision_phid: The PHID of the revision that we requested sec-approval on.
+        transactions: A list Phabricator transaction data results related to the
+            sec-approval event that we want to save.
+
+    Returns:
+        The `SecApprovalRequestEvent` that was saved to the database.
+    """
+    possible_comment_phids = []
+    for transaction in transactions:
+        phid = PhabricatorClient.expect(transaction, "phid")
+        possible_comment_phids.append(phid)
+
+    event = SecApprovalRequestEvent(
+        revision_phid=revision_phid,
+        diff_phid=diff_phid,
+        comment_candidates=possible_comment_phids,
+    )
+    db.session.add(event)
+    db.session.commit()
```

Comment:
It's best to keep use of the db session separate from the code creating and manipulating database objects. There were some race conditions in the past that required a bunch of refactoring to fix because of session use similar to this. In general, maybe a class method where you pass the information and it constructs the object would be better, and then the session management can happen right in the actual api endpoint function. For a decent explanation of the benefits of structuring things this way take a look at https://docs.sqlalchemy.org/en/13/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it